File: zig/linked-list/linked_list.zig
const std = @import("std");

pub fn LinkedList(comptime T: type) type {
    return struct {
        pub const Node = struct {
            prev: ?*Node = null,
            next: ?*Node = null,
            data: T,
        };

        first: ?*Node = null,
        last: ?*Node = null,
        len: usize = 0,

        pub fn push(self: *LinkedList(T), node: *Node) void {
            if (self.last) |last| {
                last.next = node;
                node.prev = last;
            } else {
                self.first = node;
            }

            self.last = node;
            self.len += 1;
        }

        pub fn pop(self: *LinkedList(T)) ?*Node {
            if (self.last) |last| {
                if (last == self.first) {
                    self.first = null;
                    self.last = null;
                } else {
                    if (last.prev) |prev| prev.next = null;

                    self.last = last.prev;
                    last.prev = null;
                }

                self.len -= 1;

                return last;
            } else {
                return null;
            }
        }

        pub fn shift(self: *LinkedList(T)) ?*Node {
            if (self.first) |first| {
                if (first == self.last) {
                    self.first = null;
                    self.last = null;
                } else {
                    if (first.next) |next| next.prev = null;

                    self.first = first.next;
                    first.next = null;
                }

                self.len -= 1;

                return first;
            } else {
                return null;
            }
        }

        pub fn unshift(self: *LinkedList(T), node: *Node) void {
            if (self.first) |first| {
                first.prev = node;
                node.next = first;
            } else {
                self.last = node;
            }

            self.first = node;
            self.len += 1;
        }

        pub fn delete(self: *LinkedList(T), node: *Node) void {
            var current = self.first;

            while (current) |n| : (current = n.next) {
                if (n == node) {
                    if (n.prev) |prev| prev.next = n.next;
                    if (n.next) |next| next.prev = n.prev;
                    if (n == self.first) self.first = n.next;
                    if (n == self.last) self.last = n.prev;
                    self.len -= 1;
                    break;
                }
            }
        }
    };
}

File: c/linked-list/linked_list.c
#include "linked_list.h"
#include <stdio.h>
#include <stdlib.h>

struct list_node {
  struct list_node *prev, *next;
  ll_data_t data;
};

struct list {
  struct list_node *first, *last;
};

struct list *list_create(void) { return calloc(1, sizeof(struct list)); }

size_t list_count(const struct list *list) {
  size_t count = 0;

  struct list_node *current = list->first;
  while (current != NULL) {
    current = current->next;
    count++;
  }

  return count;
}

void list_push(struct list *list, ll_data_t item_data) {
  struct list_node *new_node = calloc(1, sizeof(struct list_node));
  new_node->data = item_data;

  if (list->last == NULL) {
    list->first = new_node;
    list->last = new_node;
    return;
  }

  new_node->prev = list->last;
  list->last->next = new_node;
  list->last = new_node;
}

ll_data_t list_pop(struct list *list) {
  struct list_node *last_node = list->last;
  struct list_node *before_last_node = last_node->prev;

  list->last = before_last_node;

  if (before_last_node == NULL) {
    list->first = NULL;
  } else {
    before_last_node->next = NULL;
  }

  ll_data_t data = last_node->data;

  free(last_node);

  return data;
}

void list_unshift(struct list *list, ll_data_t item_data) {
  struct list_node *new_node = calloc(1, sizeof(struct list_node));
  new_node->data = item_data;

  if (list->first == NULL) {
    list->first = new_node;
    list->last = new_node;
    return;
  }

  new_node->next = list->first;
  list->first->prev = new_node;
  list->first = new_node;
}

ll_data_t list_shift(struct list *list) {
  struct list_node *first_node = list->first;
  struct list_node *second_node = first_node->next;

  list->first = second_node;

  if (second_node == NULL) {
    list->last = NULL;
  } else {
    second_node->prev = NULL;
  }

  ll_data_t data = first_node->data;

  free(first_node);

  return data;
}

void list_delete(struct list *list, ll_data_t data) {
  struct list_node *current = list->first;

  while (current != NULL) {
    if (current->data == data) {
      struct list_node *prev = current->prev;
      struct list_node *next = current->next;

      if (prev != NULL) {
        prev->next = current->next;
      }

      if (next != NULL) {
        next->prev = current->prev;
      }

      if (list->first == current) {
        list->first = current->next;
      }

      if (list->last == current) {
        list->last = current->prev;
      }
      return;
    }

    current = current->next;
  }
}

void list_destroy(struct list *list) {
  while (list->first != NULL) {
    struct list_node *head = list->first;
    list->first = head->next;
    free(head);
  }

  free(list);
}
